### PR TITLE
Feat/ssset should get state updates from servers and volumes

### PR DESCRIPTION
**Pull request by cristiGuranIonos**
### Description of your changes
When the ssset needs to update it should reflect this in it's state/status. This means that it needs to get updates from servers and volumes.

This PR adds:

## Checklist

I have:

- [ ] Add PR name as appropriate (e.g. `feat`/`fix`/`doc`/`test`/`refactor`)
- [ ] Run `make reviewable` and `make crds.clean` to ensure the PR is ready for review
- [ ] Add or update tests (if applicable)
- [ ] Add or update Documentation using `make docs.update` (if applicable)
- [ ] Update `docs/CHANGELOG.md` file (label: `upcoming release`)
- [ ] Check Sonar Cloud Scan
- [ ] Update Github or Jira Issue

### CODE REVIEW

File: internal/controller/compute/statefulserverset/statefulserverset_test.go [hunk 1]
```diff
@@ -453,6 +453,7 @@ func Test_areDataVolumesUpToDate(t *testing.T) {
 	type want = struct {
 		creationUpToDate bool
 		areUpToDate      bool
+		areAvailable     bool
```
Comment (fghimpeteanu-ionos) [inline internal/controller/compute/statefulserverset/statefulserverset_test.go:456]: Please add test scenarios for this new boolean. I tried to modify the value to true and there was not test failures:
image

File: internal/controller/serverset/serverset.go [hunk 2]
```diff
@@ -116,14 +116,14 @@ func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
...
-	areServersUpToDate := AreServersUpToDate(cr.Spec.ForProvider.Template.Spec, servers)
+	areServersUpToDate, areServersAvailable := AreServersReady(cr.Spec.ForProvider.Template.Spec, servers)
 
 	volumes, err := GetVolumesOfSSet(ctx, e.kube, cr.Name)
 	if err != nil {
 		return managed.ExternalObservation{}, err
 	}
 	areBootVolumesCreated := len(volumes) == cr.Spec.ForProvider.Replicas
-	areBootVolumesUpToDate := AreVolumesUpToDate(cr.Spec.ForProvider.BootVolumeTemplate, volumes)
+	areBootVolumesUpToDate, areBootVolumesAvailable := AreVolumesReady(cr.Spec.ForProvider.BootVolumeTemplate, volumes)
```
Comment (fghimpeteanu-ionos) [inline internal/controller/serverset/serverset.go:126]: Don't we need to check if NICs are available too, in the same matter we did for Servers and Volumes?

Comment (cristiGuranIonos) [reply internal/controller/serverset/serverset.go:126]: I'd like to do it in a different PR, to not have too much in this

File: internal/controller/serverset/serverset.go [hunk 3]
```diff
@@ -467,38 +471,44 @@ func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
...
 	}
 
-	return true
+	return true, true
 }
 
-// AreVolumesUpToDate checks if template params are equal to volume obj params
-func AreVolumesUpToDate(templateParams v1alpha1.BootVolumeTemplate, volumes []v1alpha1.Volume) bool {
+// AreVolumesReady checks if template params are equal to volume obj params
+func AreVolumesReady(templateParams v1alpha1.BootVolumeTemplate, volumes []v1alpha1.Volume) (bool, bool) {
```
Comment (fghimpeteanu-ionos) [inline internal/controller/serverset/serverset.go:495]: Maybe we could say here `AreBootVolumesReady` so that we know it is about the boot volumes only?

File: internal/controller/compute/statefulserverset/statefulserverset.go [hunk 4]
```diff
@@ -368,69 +374,67 @@ func isALanFieldNotUpToDate(specLan v1alpha1.StatefulServerSetLan, gotLan v1alph
...
+				return true, true, false
 			}
 		}
 	}
-	return true, true
+	return true, true, true
 }
 
 func isAVolumeFieldNotUpToDate(specVolume v1alpha1.StatefulServerSetVolume, volumeIndex int, volumes []v1alpha1.Volume) bool {
-	if generateProviderNameFromIndex(specVolume.Metadata.Name, volumeIndex) != volumes[volumeIndex].Spec.ForProvider.Name {
```
Comment (fghimpeteanu-ionos) [inline internal/controller/compute/statefulserverset/statefulserverset.go:387]: Why was this removed? Was it because it did not match the way we create the name of the volume (e.g. `<nameFromSpec-replicaIdx-volumeIdx>`)
image
Shouldn't we in this case just update it to what it should be instead of removing it?

**Review by fghimpeteanu-ionos — APPROVED** [review]

File: internal/controller/compute/statefulserverset/statefulserverset.go (hunk 4 above)
Comment (cristiGuranIonos) [reply internal/controller/compute/statefulserverset/statefulserverset.go:387]: changing metadata name means a new object is created. We could support metadata name update, but it gets tricky. We should actually just make them all immutable for now.
